scheduler/src/cook/rest/api.clj

```diff
@@ -1969,6 +1973,50 @@
       (seq errors) [false {::error (str/join "\n" errors)}]
       :else true)))
 
+(defn user-queue-length-within-limit?
+  "Check if the job submission would cause the user to have more jobs queued than they're
+  allowed. We refresh queue lengths from the database on a configurable interval. Note that
+  this check does not take into account any of the following in real-time, and therefore is
+  only 100% accurate as of the last refresh from the database:
+
+  - Jobs will get removed from the queue when they launch
+  - Jobs will get added to the queue if they fail and have retries
+  - If there are multiple API hosts in the cluster, each one is maintaining a separate
+    count of the queue, without knowing about the other API hosts' job submissions
+
```

Comment:
A suggestion to improve this is to give explicit error bars in the explanation.
"""
There are some errors in this:
   -- Jobs will get removed from teh queue then they launch. However, We limit users to 3-20 jobs/sec, so in 2 minutes a user can launch at most 200-2k jobs. 
   -- Jobs will get added to the queue if they are running, fail and have retries. Again, this is limited based on how many jobs a user has running.
   --- If there are multiple API hosts... This only lets a user go over-queue.
   --- If a user kills their jobs in the queue, they may be unable to reuse that for 2 minutes. To mitigiate this ....

However, we intend per-user queue limits to be in the 10k-50k range, so the errors are fairly minor, except for the kill jobs case.
"""

(We need to both describe and handle the kill jobs case IMO. My suggestion is hope for API server locality and decrement the queue on kill for a job that's still alive.)

Reply:
Good call on killing jobs, I've now added that.

The error bars are dependent on the configuration you run Cook with, so I don't think we should give specific numbers about them in the code.

Reply:
Good call, done